### PR TITLE
Fix rrdir AsyncIterable + clean up picomatch

### PR DESCRIPTION
**Pull request by grant-dennison**
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.) *only directly using `rrdir` (which depends on `picomatch`)*
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).
  - [x] `npm test rrdir`
  - [x] `npm test picomatch`
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: (see various links in description of changes below)
- [x] ~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~

The driving motivation for this pull request is that I wanted to use the `rrdir` package but found that the first example in the [README](https://www.npmjs.com/package/rrdir) did not work because the types in `@types/rrdir` were incorrect. Specifically, this example did not work:
```typescript
for await (const entry of rrdir("dir")) {
  // => {path: 'dir/file', directory: false, symlink: false}
}
```
When I looked at the source for the type definitions, I found that `rrdir()` and `rrdir.sync()` both had a return type of `Promise<entry[]>` when that type is only applicable to `rrdir.sync()`.

I used the [`rrdir` usage snippets](https://github.com/silverwind/rrdir#usage) as literally as possible in the new tests to verify the types.

I got a little carried away and wound up doing a bunch of cleanup alongside this core change.
- I realized that [`options.match`](https://github.com/silverwind/rrdir#options-object) corresponded to [`picomatch`](https://www.npmjs.com/package/picomatch), so I wanted to **get rid of that foreign type from `@types/rrdir`**.
- When I found that the options type was not exported from the `picomatch` types package, I had to **split the files to allow export** of the options type for use in `rrdir`. *(For any reviewers, note that I did try to separate the commit where I split the files from the commit where I changed the types.)*
- I also found that the options type in `@types/picomatch` did not fully reflect the [documented options](https://www.npmjs.com/package/picomatch#options), but the corresponding type in `@types/rrdir` was more complete. I spent some time **reconciling the two to match the documentation** as best I could.
- I went ahead and **copy-pasted the documentation** from the README for each `picomatch` option.
- **Adding some tests** to `picomatch` (there were none), I found some more inconsistencies with the README as I tried to directly use the example code snippets.
  - **`Picomatch#test()` returned the wrong type**. I dug up the correct types from [the source].(https://github.com/micromatch/picomatch/blob/master/lib/picomatch.js#L117).
  - `Picomatch#matchBase()` did not have **options [optional as documented](https://www.npmjs.com/package/picomatch#matchbase)**.
- I also added the [`Buffer` version of `rrdir` API](https://github.com/silverwind/rrdir#dir-string--buffer) since I noticed it was missing.

### CODE REVIEW

**Comment by grant-dennison** [comment]
@p-kuen I believe you are the "@Patcher56" the bot tried to ping. Please see the [comment above](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51523#issuecomment-789423275) requesting your review.

**Review by p-kuen — COMMENTED** [review]
Thank you!
What do you exactly mean by "automatic exporting"?

**Comment by grant-dennison** [comment]
> What do you exactly mean by "automatic exporting"?

@p-kuen I changed the `export =` to a more typical (as it seems to me) simple export (`export interface ...`). When I did that, the linter gave me some errors about the file automatically exporting everything. The linter suggested that I could disable automatic exporting for the file by using `export {}`.

I did a little bit more digging and found [a related discussion](https://github.com/Microsoft/dtslint/issues/138#issue-353246624) about hiding some types in .d.ts files.

**Comment by peterblazejewicz** [comment]
I think definition has a room to improve, including removing files that are really not to be seen in the original JS source. The `picomatch` seems to be a set of functional CommonJS modules, exported via picomatch re-export for consumption.

I'd start with (using separate, not inlined module definitions):

```ts
// Type definitions for picomatch 2.2
// Project: https://github.com/micromatch/picomatch
// Definitions by: Patrick <https://github.com/Patcher56>
// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped

 import picomatch = require('./lib/picomatch');

 export = picomatch;
 ```
and for example starting here (still to be added imports from `/lib/contants.d.ts`, etc)
`./lib/picomatch.d.ts`

```ts
/**
 * Creates a matcher function from one or more glob patterns. The
 * returned function takes a string to match as its first argument,
 * and returns true if the string is a match. The returned matcher
 * function also takes a boolean as the second argument that, when true,
 * returns an object with additional information.
 *
 * ```js
 * const picomatch = require('picomatch');
 * // picomatch(glob[, options]);
 *
 * const isMatch = picomatch('*.!(*a)');
 * console.log(isMatch('a.a')); //=> false
 * console.log(isMatch('a.b')); //=> true
 * ```
 * @param glob One or more glob patterns.
 * @return Returns a matcher function.
 * @api public
 */
declare function picomatch(
    glob: string | string[],
    options?: picomatch.PicomatchOptions,
    returnState?: boolean,
): picomatch.Matcher;

declare namespace picomatch {
    interface Matcher {
        (test: string): boolean;
    }
    interface PicomatchOptions {
        /**
         * If set, then patterns without slashes will be matched against the basename of the path if it contains slashes.
         * For example, `a?b` would match the path `/xyz/123/acb`, but not `/xyz/acb/123`.
         */
        basename?: boolean;
        /**
         * Follow bash matching rules more strictly - disallows backslashes as escape characters, and treats single stars as globstars (`**`).
         */
        bash?: boolean;
        /**
         * Return regex matches in supporting methods.
         */
        capture?: boolean;
        /**
         * Allows glob to match any part of the given string(s).
         */
        contains?: boolean;
        /**
         * Current working directory. Used by `picomatch.split()`
         */
        cwd?: string;
        /**
         * Debug regular expressions when an error is thrown.
         */
        debug?: boolean;
        /**
         * Enable dotfile matching. By default, dotfiles are ignored unless a `.` is explicitly defined in the pattern, or `options.dot` is true
         */
        dot?: boolean;
        /**
         * Custom function for expanding ranges in brace patterns, such as `{a..z}`.
         * The function receives the range values as two arguments, and it must return a string to be used in the generated regex.
         * It's recommended that returned strings be wrapped in parentheses.
         */
        expandRange?: (a: string, b: string) => string;
        /**
         * Throws an error if no matches are found. Based on the bash option of the same name.
         */
        failglob?: boolean;
        /**
         * To speed up processing, full parsing is skipped for a handful common glob patterns. Disable this behavior by setting this option to `false`.
         */
        fastpaths?: boolean;
        /**
         * Regex flags to use in the generated regex. If defined, the `nocase` option will be overridden.
         */
        flags?: boolean;
        /**
         * Custom function for formatting the returned string. This is useful for removing leading slashes, converting Windows paths to Posix paths, etc.
         */
        format?: (str: string) => string;
        /**
         * One or more glob patterns for excluding strings that should not be matched from the result.
         */
        ignore?: string | string[];
        /**
         * Retain quotes in the generated regex, since quotes may also be used as an alternative to backslashes.
         */
        keepQuotes?: boolean;
        /**
         * When `true`, brackets in the glob pattern will be escaped so that only literal brackets will be matched.
         */
        literalBrackets?: boolean;
        /**
         * Support regex positive and negative lookbehinds. Note that you must be using Node 8.1.10 or higher to enable regex lookbehinds.
         */
        lookbehinds?: boolean;
        /**
         * Alias for `basename`
         */
        matchBase?: boolean;
        /**
         * Limit the max length of the input string. An error is thrown if the input string is longer than this value.
         */
        maxLength?: boolean;
        /**
         * Disable brace matching, so that `{a,b}` and `{1..3}` would be treated as literal characters.
         */
        nobrace?: boolean;
        /**
         * Disable brace matching, so that `{a,b}` and `{1..3}` would be treated as literal characters.
         */
        nobracket?: boolean;
        /**
         * Make matching case-insensitive. Equivalent to the regex `i` flag. Note that this option is overridden by the `flags` option.
         */
        nocase?: boolean;
        /**
         * @deprecated use `nounique` instead.
         * This option will be removed in a future major release. By default duplicates are removed.
         * Disable uniquification by setting this option to false.
         */
        nodupes?: boolean;
        /**
         * Alias for `noextglob`
         */
        noext?: boolean;
        /**
         * Disable support for matching with extglobs (like `+(a\|b)`)
         */
        noextglob?: boolean;
        /**
         * Disable support for matching nested directories with globstars (`**`)
         */
        noglobstar?: boolean;
        /**
         * Disable support for negating with leading `!`
         */
        nonegate?: boolean;
        /**
         * Disable support for regex quantifiers (like `a{1,2}`) and treat them as brace patterns to be expanded.
         */
        noquantifiers?: boolean;
        /**
         * Function to be called on ignored items.
         */
        onIgnore?: (result: Result) => void;
        /**
         * Function to be called on matched items.
         */
        onMatch?: (result: Result) => void;
        /**
         * Function to be called on all items, regardless of whether or not they are matched or ignored.
         */
        onResult?: (result: Result) => void;
        /**
         * Support POSIX character classes ("posix brackets").
         */
        posix?: boolean;
        /**
         * Convert all slashes in file paths to forward slashes. This does not convert slashes in the glob pattern itself
         */
        posixSlashes?: boolean;
        /**
         * Convert all slashes in file paths to forward slashes. This does not convert slashes in the glob pattern itself
         */
        prepend?: boolean;
        /**
         * Use regular expression rules for `+` (instead of matching literal `+`), and for stars that follow closing parentheses or brackets (as in `)*` and `]*`).
         */
        regex?: boolean;
        /**
         * Throw an error if brackets, braces, or parens are imbalanced.
         */
        strictBrackets?: boolean;
        /**
         * When true, picomatch won't match trailing slashes with single stars.
         */
        strictSlashes?: boolean;
        /**
         * Remove backslashes preceding escaped characters in the glob pattern. By default, backslashes are retained.
         */
        unescape?: boolean;
        /**
         * Alias for `posixSlashes`, for backwards compatibility.
         */
        unixify?: boolean;
    }

    function test(
        input: string,
        regex: RegExp,
        options?: PicomatchOptions,
        test?: {},
    ): { isMatch: boolean; match?: boolean | RegExpExecArray | null; output: string };

    function matchBase(input: string, glob: RegExp | string, options?: {}, posix?: any): boolean;

    function isMatch(str: string | string[], patterns: string | string[], options?: {}): boolean;

    function parse(pattern: string, options: {}): {};

    function scan(input: string, options: {}): {};
}

export = picomatch;
```

Just with that definition above tests are still passing:

```ts
import pm = require('picomatch');
// just to show we can import types
import type { Matcher, PicomatchOptions } from 'picomatch';

// main function
const isMatch = pm('*.!(*a)');

// $ExpectType boolean
isMatch('abcd');

const testResult = pm.test('foo/bar', /^(?:([^/]*?)\/([^/]*?))$/);
// $ExpectType boolean
testResult.isMatch;
// $ExpectType string
testResult.output;

// $ExpectType boolean
pm.matchBase('foo/bar.js', '*.js');
```

**Comment by grant-dennison** [comment]
@peterblazejewicz Thank you for reviewing this! I'm trying really hard to understand what you are saying, but I don't think I'm fully grasping the significance. It seems like you are making a few different points, but I don't think I understand them all. Before I make any changes, I want to clarify some things:

1. I think you are saying that the .d.ts files here in DefinitelyTyped should try to match one-to-one with files in source rather than just being functionally equivalent as I tried to do. Am I understanding that part right?
2. It seems like you are advocating the use of `require()`, `declare`, and `namespace` over just `import` and `export`. My understanding was that TypeScript officially recommends [always using modules](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#using-modules) over namespaces. I've also gotten the impression (though I can't find a source) that `import` is always preferred to `require` in TypeScript. Is there some kind of necessary condition here for using `require()` maybe as briefly touched [in TypeScript docs](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)?
3. It looks like you left out a couple of functions in your sample. Unless you tell me otherwise, I'll assume this was an oversight and not an implicit suggestion to remove those parts of the API.

**Comment by peterblazejewicz** [comment]
> 1. I think you are saying that the .d.ts files here in DefinitelyTyped should try to match one-to-one with files in source rather than just being functionally equivalent as I tried to do. Am I understanding that part right?

yes, because this is open project, so would be best, if a folk like me can see the same concepts (names, classes, literals, etc) in DT package, compared to JS source. Literal, would be the best, it's easy to understand.

> 2. It seems like you are advocating the use of require(), declare, and namespace over just import and export. My understanding was that TypeScript officially recommends always using modules over namespaces. I've also gotten the impression (though I can't find a source) that import is always preferred to require in TypeScript. Is there some kind of necessary condition here for using require() maybe as briefly touched in TypeScript docs?

Oh, everything is in proper place. Those are for consuming TS files. Here, on DT, we promote using a solution that just works best accross different TS versions (no matter how the code is later consumed).
So relevant doc would be:

> This older syntax is harder to use but works everywhere. 

https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-d-ts.html#default-exports



> 3\. It looks like you left out a couple of functions in your sample. Unless you tell me otherwise, I'll assume this was an oversight and not an implicit suggestion to remove those parts of the API.

Yes, it was only part required to existing tests to pass again and worked as example.

Before commiting to anything, midn, it was just my comment on the package itself, not your PR. It just happens, in your PR there is a log of stuff shuffled wtihin a package itself, so I though I would drop my 2c, thx!

**Comment by grant-dennison** [comment]
@peterblazejewicz I implemented some changes that I think are mostly in line with your suggestions. The main thing I did **not** do was pull `PicomatchOptions` into the main `picomatch.d.ts` file.

Also:
- Added more half-baked (without `$ExpectType` in most cases) tests from snippets directly in picomatch README.
- Fixed some more problems with the API I found while digging through the picomatch source. (I definitely didn't thoroughly complete this. For example, `parse()` is still missing a bunch of stuff. But I'm not willing to tackle that right now.)

File: types/picomatch/lib/picomatch-options.d.ts
Comment (p-kuen) [inline types/picomatch/lib/picomatch-options.d.ts:175]: Should be removed IMHO

File: types/picomatch/index.d.ts
Comment (p-kuen) [inline types/picomatch/index.d.ts:7]: Why do we need everything in a lib folder?

File: types/picomatch/lib/picomatch.d.ts
Comment (p-kuen) [inline types/picomatch/lib/picomatch.d.ts:4]: We do we need to use as here?

File: types/picomatch/lib/picomatch.d.ts
Comment (p-kuen) [inline types/picomatch/lib/picomatch.d.ts:27]: Just use PicomatchOptions here

File: types/picomatch/lib/picomatch.d.ts
Comment (p-kuen) [inline types/picomatch/lib/picomatch.d.ts:36]: Not required IMHO

**Review by p-kuen — COMMENTED** [review]
It seems like we are mixing commonjs and es6 imports now. Not sure if this is good.

File: types/picomatch/index.d.ts
Comment (grant-dennison) [reply types/picomatch/index.d.ts:7]: I moved them there per [Piotr's comment](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/51523#issuecomment-803549366) because the source files in picomatch itself are [in the lib folder](https://github.com/micromatch/picomatch/tree/master/lib).

File: types/picomatch/lib/picomatch.d.ts
Comment (grant-dennison) [reply types/picomatch/lib/picomatch.d.ts:4]: There is aliasing going on when we re-expose `PicomatchOptions` in the exposed namespace. `type PicomatchOptions = PicomatchOptions;` would be no bueno.

File: types/picomatch/lib/picomatch.d.ts
Comment (grant-dennison) [reply types/picomatch/lib/picomatch.d.ts:27]: Plain `PicomatchOptions` isn't available here as is. My choices with the current implementation are `PicomatchOptionsImport` or `picomatch.PicomatchOptions`.

File: types/picomatch/lib/picomatch.d.ts
Comment (grant-dennison) [reply types/picomatch/lib/picomatch.d.ts:36]: In the part of the review that you may not have been looking at, I'm using this type in the `rrdir` .d.ts files. This was really the reason I touched anything at all in `picomatch` in the first place. On this second round (after review) it seemed to me like it would be better to expose this type through the single point of export rather than manually grabbing the file containing this options type. I suspect that other libraries may also want to have access to this type as I did in `rrdir` since the implementation there was tied to `picomatch`.

File: types/picomatch/lib/picomatch-options.d.ts
Comment (grant-dennison) [reply types/picomatch/lib/picomatch-options.d.ts:175]: If I remove this and nothing else, it makes dtslint unhappy. I might be able to rework this file not to use es6 stuff and then remove this, but I do like that this file doesn't imply it has anything but types in it. There is no corresponding file for this in picomatch source, so people would ideally only `import type` from this so no 'require()'s wind up in the compiled output.

**Comment by grant-dennison** [comment]
@p-kuen WRT mixing commonjs and es6, I'm not sure what to say. I tried to go more pure es6, but I shifted direction per Piotr's review. I think I will wait for Piotr to weigh in before I do anything further with this.

Thanks for coming back to review this again!

**Comment by peterblazejewicz** [comment]
@grant-dennison let's stay with your current changes Once done and merges, we an pick topic back. Sorry, if I introduced a confusion.

File: types/picomatch/lib/picomatch-options.d.ts
Comment (andrewbranch) [inline types/picomatch/lib/picomatch-options.d.ts:3]: Making this a separate file _does_ imply that `node_modules/picomatch/lib/picomatch-options.js` exists, though it’s not so bad since you’re not implying that any values are exported from it. Nonetheless, this could theoretically cause issues. Unfortunately in DT we aren’t at the liberty to organize files however we see fit; we have to keep a 1:1 mapping to the JS.

File: types/picomatch/lib/scan.d.ts
Comment (andrewbranch) [inline types/picomatch/lib/scan.d.ts:1]: It would be nice to bundle these auxiliary types into a `namespace scan` so they can be exported alongside the function.

File: types/rrdir/index.d.ts
Comment (andrewbranch) [inline types/rrdir/index.d.ts:35]: Maybe

```suggestion
interface Entry<T extends string | Buffer = string> {
```

File: types/rrdir/index.d.ts
Comment (andrewbranch) [inline types/rrdir/index.d.ts:43]: I’d probably get rid of these and write them inline

**Review by andrewbranch — CHANGES_REQUESTED** [review]

**Comment by andrewbranch** [comment]
My feedback is addressed. Leaning on @p-kuen and @peterblazejewicz for the substantive changes—does it look ok to you?

**Comment by grant-dennison** [comment]
I'm a bit miffed that [PR 52329](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52329) made it through review within a day, but this PR hasn't gone through in over a month (even though I've been on top of it). @p-kuen @peterblazejewicz @andrewbranch what can I do to push this along? I've already been through two rounds of code review and am about to now deal with a first set of conflict resolutions.

**Comment by p-kuen** [comment]
@grant-dennison please be aware that this PR is a huge change and needs to be reviewed a lot longer than the mentioned PR. We all have a lot to do beside reviewing some open source packages.

The other thing is that your changes are not split up in multiple PRs. We have file changes, syntax changes, performance changes, etc. here all at once, which is not easy to review. So please be patient, we will review it soon.

Thanks.

File: types/picomatch/lib/picomatch.d.ts
Comment (rbuckton) [inline types/picomatch/lib/picomatch.d.ts:37]: This lets you remove the duplicate import for `./parse`, above.

```suggestion
        state: parseImport.ParseState;
```

File: types/picomatch/lib/picomatch.d.ts
Comment (rbuckton) [inline types/picomatch/lib/picomatch.d.ts:232]:
```suggestion
        state: parseImport.ParseState,
```

File: types/rrdir/index.d.ts
Comment (rbuckton) [inline types/rrdir/index.d.ts:6]:
```suggestion
```

This reference comment means that `rrdir` *introduces* `esnext.asynciterable` (even if its not specified by the user), not that it *depends* on it. You already have `/// <reference types="node" />` below, which *does* introduce `esnext.asynciterable` (and already has a lib reference for it), so this line shouldn't be needed.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/4486794741ccb9f18f562fefade043ac31e817de/types/node/base.d.ts#L11

**Review by rbuckton — CHANGES_REQUESTED** [review]

File: types/rrdir/index.d.ts
Comment (grant-dennison) [reply types/rrdir/index.d.ts:6]: Ok. Do you have a link handy to some article that might talk about this more in depth? This was one of those pieces I did not understand well.

**Review by orta — COMMENTED** [review]
This all looks pretty reasonable to me, kinda just awaiting someone knowing the domain a bit better 👍🏻

**Comment by elibarzilay** [comment]
(Looks like this still has a bunch of unresolved comments that need to be addressed?)

**Comment by grant-dennison** [comment]
@elibarzilay To my knowledge, I have addressed all of the comments. Most of them I have made changes in response to, but some of them I just commented in response. For all of the unresolved comments, I am waiting for the commentator to either resolve the comment or comment back why it is still unresolved.

**Comment by grant-dennison** [comment]
Thanks @elibarzilay. Sorry about the mess. From my perspective, it's dangerous for the author to resolve the threads (started by reviewers) themselves because it doesn't guarantee somebody double-checked their work (i.e. the author could effectively just cover up review comments they don't like). But if that's the review pattern for this repo, I'll try to follow that if I contribute here again.

**Comment by elibarzilay** [comment]
> Thanks @elibarzilay. Sorry about the mess. From my perspective, it's dangerous for the author to resolve the threads (started by reviewers) themselves because it doesn't guarantee somebody double-checked their work (i.e. the author could effectively just cover up review comments they don't like). But if that's the review pattern for this repo, I'll try to follow that if I contribute here again.

Normally you'd be right, but DT is different in the sheer volume of work that goes through it, so maintainer attention is very scarce, and even people who review your PR would generally not spend much time verifying changes etc.

**Pull request merged**
